### PR TITLE
"As of" feature

### PR DESCRIPTION
**Pull request by kateharwood**

### CODE REVIEW

File: dashboard/app.R [hunk 1]
```diff
@@ -334,30 +336,53 @@ server <- function(input, output, session) {
         locationSubtitleText = paste0(', Location: ', aggregate ,' over all states and territories common to these forecasters*')
         output$renderLocations <- renderText(toupper(locationsIntersect))
         output$renderWarningText = renderText("")
+        showElement("truthPlot")
```
Comment (nmdefries) [inline dashboard/app.R:339]: It would be nice for the truth plot to have a legend for the as-of vs most recent data.

**Review by nmdefries — CHANGES_REQUESTED** [review]

File: dashboard/app.R [hunk 2]
```diff
@@ -8,11 +8,17 @@ library(plotly)
...
 source('./common.R')
 
 # All data is fully loaded from AWS
-dataLoaded = FALSE
+DATA_LOADED = FALSE
+
+# Earliest 'as of' date available from covidcast API
+MIN_AVAIL_NATION_AS_OF_DATE = as.Date('2021-01-09')
+MIN_AVAIL_HOSP_AS_OF_DATE  = as.Date('2021-01-13')
+MIN_AVAIL_TERRITORY_AS_OF_DATE = as.Date('2021-02-13')
```
[the remark this answers is not in the document]
Comment (krivard) [reply dashboard/app.R:21]: `covidcast_meta()` will give you this information too, though if the additional load time of the extra API call is prohibitive i'm okay with hard-coding them

File: dashboard/app.R [hunk 3]
```diff
@@ -262,55 +278,41 @@ server <- function(input, output, session) {
...
-      plotTitle = "Coverage"
+    if (!is.null(asOfData) && dim(asOfData)[1] != 0) {
+      asOfData = asOfData %>% rename(target_end_date = time_value, as_of_actual = value)
+      asOfData = asOfData[c("target_end_date", "geo_value", "as_of_actual")]
+
+      # Since cases and deaths are shown as weekly incidence, but the "as of" data from the covidcast API
+      # is daily, we need to sum over the days leading up to the target_end_date of each week to get the
+      # weekly incidence.
+      # Get the 'as of' dates that are the target_end_dates in the scoring df
+      dateGroupDf = asOfData %>% filter(asOfData$target_end_date %in% filteredScoreDf$target_end_date)
```
Comment (krivard) [inline dashboard/app.R:301]: what will happen if asOfData has records for some of the 6 days leading up to a date listed in filteredScoreDf$target_end_date, but no records for dates matching it? eg if filteredScoreDf$target_end_date has 20200107, and asOfData$target_end_date has 20200101, 20200102, 20200103, 20200104, 20200105, and 20200106, but not 20200107?

File: dashboard/app.R [hunk 4]
```diff
@@ -262,55 +278,41 @@ server <- function(input, output, session) {
...
+      if (dim(dateGroupDf)[1] != 0) {
+        asOfData = filterAsOfData(asOfData, dateGroupDf, filteredScoreDf)
```
Comment (krivard) [inline dashboard/app.R:303]:
```suggestion
        # sum over preceding week for all weekly target variables
        asOfData = filterAsOfData(asOfData, dateGroupDf, filteredScoreDf)
```

File: dashboard/app.R [hunk 5]
```diff
@@ -375,11 +400,33 @@ server <- function(input, output, session) {
...
     titleText = paste0('<b>', plotTitle,'</b>','<br>', '<sup>',
                        'Target Variable: ', input$targetVariable,
                        locationSubtitleText, '<br>',
                        tags$span(id="drag-to-zoom", " Drag to zoom"),
                        '</sup>')
+
+    # Some forecasters have multiple forecasts for the same week end date with the same score,
+    # and tsibbles needs unique keys/indexes
+    # So we remove dups before setting tsibble
+    # TODO this "fix" messes up the faceted plots, need to figure this out
```
Comment (krivard) [inline dashboard/app.R:426]: is this a blocking TODO? if not dump it in an issue and cite the link here

File: dashboard/app.R [hunk 6]
```diff
@@ -582,6 +732,85 @@ server <- function(input, output, session) {
...
   })
 
+  updateAsOfData = function() {
+    dataSource = "jhu-csse"
+    if(input$targetVariable == "Cases") {
+      targetSignal = "confirmed_incidence_num"
+    } else if (input$targetVariable == "Deaths") {
+      targetSignal = "deaths_incidence_num"
+    } else if (input$targetVariable == "Hospitalizations") {
+      targetSignal = "confirmed_admissions_covid_1d"
```
Comment (krivard) [inline dashboard/app.R:742]: @brookslogan just double checking that HUB is forecasting confirmed adult and pediatric admissions (=the signal used here), and not confirmed+suspected

File: dashboard/app.R (hunk 1 above)
Comment (kateharwood) [reply dashboard/app.R:339]: That is coming in the next PR adding forecasts

File: dashboard/app.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (kateharwood) [reply dashboard/app.R]: Thanks for the heads up. I've updated these min dates to be more nuanced. Based on the data coming in from the API, the hospitalization data seems to have started around 2020-11-11 for both states and nation (or at least nothing is coming back before then for either).

The state data for cases and deaths goes way back, but nation cases and deaths data only comes back for dates in 2021.

Note: for deaths and cases, when using the "coverage" scoring or when totaling over all locations, we must go with the territories min date bc territories are included in those calculations

File: dashboard/app.R (hunk 3 above)
Comment (kateharwood) [reply dashboard/app.R:298]: If I understand what you're getting at, I think this line would take care of it:

```
      # In the case where there are target week end days missing from the scoring or as of data 
      # we don't want to end up summing values over multiple weeks so we make sure each date_group only spans one week
      asOfData = asOfData %>% filter(asOfData$date_group - asOfData$target_end_date < 7)
```

File: dashboard/app.R (hunk 4 above)
Comment (kateharwood) [reply dashboard/app.R:303]: Moved the above detailed comment that explains the summing to here to be more relevant and added this comment to the line in the function that actually performs the summation.

File: dashboard/app.R (hunk 5 above)
Comment (kateharwood) [reply dashboard/app.R]: Done.

File: dashboard/app.R [hunk 7]
```diff
@@ -8,11 +8,17 @@ library(plotly)
 library(shinyjs)
 library(tsibble)
 library(aws.s3)
+library(covidcast)
```
[the remark this answers is not in the document]
Comment (kateharwood) [reply dashboard/app.R:11]: Done.

**Review by krivard — APPROVED** [review]
🚀

File: dashboard/app.R (hunk 6 above)
Comment (kateharwood) [reply dashboard/app.R:735]: Will confirm in the slack channel. Merging to dev for now, can change the signal if needed later.

**Review by nmdefries — APPROVED** [review]

**Pull request merged**